Eliminate compiler warnings and turn on -Werr

I used the master branch.

```
../modules/afprog/afprog.c: In function 'afprogram_reload_store_item_free':
../modules/afprog/afprog.c:69:3: warning: passing argument 1 of 'log_pipe_unref' from incompatible pointer type [enabled by default]
   log_pipe_unref(reload_info->writer);
   ^
In file included from ../lib/driver.h:29:0,
                 from ../modules/afprog/afprog.h:27,
                 from ../modules/afprog/afprog.c:24:
../lib/logpipe.h:237:6: note: expected 'struct LogPipe *' but argument is of type 'struct LogWriter *'
 void log_pipe_unref(LogPipe *self);
      ^
../modules/afprog/afprog.c: In function 'afprogram_dd_kill_child':
../modules/afprog/afprog.c:329:13: warning: unused variable 'pgid' [-Wunused-variable]
       pid_t pgid;
             ^
../modules/afprog/afprog.c: In function 'afprogram_dd_store_reload_store_item':
../modules/afprog/afprog.c:452:3: warning: passing argument 4 of 'cfg_persist_config_add' from incompatible pointer type [enabled by default]
   cfg_persist_config_add(cfg, afprogram_dd_format_persist_name(self), reload_info, afprogram_reload_store_item_destroy_notify, FALSE);
   ^
In file included from ../lib/logpipe.h:30:0,
                 from ../lib/driver.h:29,
                 from ../modules/afprog/afprog.h:27,
                 from ../modules/afprog/afprog.c:24:
../lib/cfg.h:158:6: note: expected 'GDestroyNotify' but argument is of type 'void (*)(struct AFProgramReloadStoreItem *)'
 void cfg_persist_config_add(GlobalConfig *cfg, gchar *name, gpointer value, GDestroyNotify destroy, gboolean force);

```

To do:
- [x] Eliminate compiler warnings - #779
- [ ] Turn on -Werr #871

I saw these too, and this is pretty unfortunate as these are recent
additions.


-- 
Bazsi

On Tue, Nov 10, 2015 at 2:05 PM, Tibor Benke <notifications@github.com>
wrote:

> ../modules/afprog/afprog.c: In function 'afprogram_reload_store_item_free':
> ../modules/afprog/afprog.c:69:3: warning: passing argument 1 of 'log_pipe_unref' from incompatible pointer type [enabled by default]
>    log_pipe_unref(reload_info->writer);
>    ^
> In file included from ../lib/driver.h:29:0,
>                  from ../modules/afprog/afprog.h:27,
>                  from ../modules/afprog/afprog.c:24:
> ../lib/logpipe.h:237:6: note: expected 'struct LogPipe *' but argument is of type 'struct LogWriter *'
>  void log_pipe_unref(LogPipe *self);
>       ^
> ../modules/afprog/afprog.c: In function 'afprogram_dd_kill_child':
> ../modules/afprog/afprog.c:329:13: warning: unused variable 'pgid' [-Wunused-variable]
>        pid_t pgid;
>              ^
> ../modules/afprog/afprog.c: In function 'afprogram_dd_store_reload_store_item':
> ../modules/afprog/afprog.c:452:3: warning: passing argument 4 of 'cfg_persist_config_add' from incompatible pointer type [enabled by default]
>    cfg_persist_config_add(cfg, afprogram_dd_format_persist_name(self), reload_info, afprogram_reload_store_item_destroy_notify, FALSE);
>    ^
> In file included from ../lib/logpipe.h:30:0,
>                  from ../lib/driver.h:29,
>                  from ../modules/afprog/afprog.h:27,
>                  from ../modules/afprog/afprog.c:24:
> ../lib/cfg.h:158:6: note: expected 'GDestroyNotify' but argument is of type 'void (*)(struct AFProgramReloadStoreItem *)'
>  void cfg_persist_config_add(GlobalConfig *cfg, gchar *name, gpointer value, GDestroyNotify destroy, gboolean force);
>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balabit/syslog-ng/issues/776>.
>

I think the OSE tree is not that far, we only have a handful of warnings,
and half of them is this set. Apart from these we have 5 warnings (2 in
testcases, 2 in the sql destination, and one in the lexer).

```
lib/cfg-lex.c:2246:13: warning: 'yy_fatal_error' defined but not used
[-Wunused-function]
 static void yy_fatal_error (yyconst char* msg , yyscan_t yyscanner)
             ^
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/modules/afsql/afsql.c:
In function 'afsql_dd_ensure_initialized_connection':
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/modules/afsql/afsql.c:705:3:
warning: 'dbi_conn_new' is deprecated (declared at
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/install/include/dbi/dbi.h:208)
[-Wdeprecated-declarations]
   self->dbi_ctx = dbi_conn_new(self->type);
   ^
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/modules/afsql/afsql.c:
In function 'afsql_dd_init':
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/modules/afsql/afsql.c:1300:7:
warning: 'dbi_initialize' is deprecated (declared at
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/install/include/dbi/dbi.h:179)
[-Wdeprecated-declarations]
       gint rc = dbi_initialize(NULL);
       ^
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/modules/afprog/afprog.c:
In function 'afprogram_reload_store_item_free':
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/modules/afprog/afprog.c:69:18:
warning: passing argument 1 of 'log_pipe_unref' from incompatible pointer
type
   log_pipe_unref(reload_info->writer);
                  ^
In file included from
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/lib/driver.h:29:0,
                 from
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/modules/afprog/afprog.h:27,
                 from
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/modules/afprog/afprog.c:24:
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/lib/logpipe.h:237:6:
note: expected 'struct LogPipe *' but argument is of type 'struct LogWriter
*'
 void log_pipe_unref(LogPipe *self);
      ^
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/modules/afprog/afprog.c:
In function 'afprogram_dd_kill_child':
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/modules/afprog/afprog.c:329:13:
warning: unused variable 'pgid' [-Wunused-variable]
       pid_t pgid;
             ^
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/modules/afprog/afprog.c:
In function 'afprogram_dd_store_reload_store_item':
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/modules/afprog/afprog.c:452:84:
warning: passing argument 4 of 'cfg_persist_config_add' from incompatible
pointer type
   cfg_persist_config_add(cfg, afprogram_dd_format_persist_name(self),
reload_info, afprogram_reload_store_item_destroy_notify, FALSE);

^
In file included from
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/lib/logpipe.h:30:0,
                 from
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/lib/driver.h:29,
                 from
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/modules/afprog/afprog.h:27,
                 from
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/modules/afprog/afprog.c:24:
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/lib/cfg.h:158:6:
note: expected 'GDestroyNotify' but argument is of type 'void (*)(struct
AFProgramReloadStoreItem *)'
 void cfg_persist_config_add(GlobalConfig *cfg, gchar *name, gpointer
value, GDestroyNotify destroy, gboolean force);
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/modules/kvformat/tests/test_kv_parser.c:81:1:
warning: ‘assert_kv_parser_fails’ defined but not used [-Wunused-function]
 assert_kv_parser_fails(const gchar *kv)
 ^
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/lib/tests/test_log_message.c:
In function
‘test_log_msg_set_value_indirect_with_self_referencing_handle_results_in_a_nonindirect_value’:
/home/bazsi/zwa/projects/syslog-ng-ose-3.6/source/syslog-ng/lib/tests/test_log_message.c:279:12:
warning: unused variable ‘handle’ [-Wunused-variable]
   NVHandle handle;
            ^
```

-- 
Bazsi

On Tue, Nov 10, 2015 at 4:28 PM, Budai Laszlo <notifications@github.com>
wrote:

> ehh... I think we should do something with these warnings :-(
> Maybe Travis/drone.io and -Werr could help ... but we are far from that.
> (or create a check-warning target in the makefile)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balabit/syslog-ng/issues/776#issuecomment-155451687>.
>

I wanted to turn this on, but I see some warnings in internal ivykis:

```
/bin/bash ../libtool --tag=CC   --mode=link gcc  -Werror   -o iv_event_test iv_event_test.o ../src/libivykis.la
libtool: link: gcc -Werror -o iv_event_test iv_event_test.o  ../src/.libs/libivykis.a
../src/.libs/libivykis.a(iv_main_posix.o): In function `__iv_deinit':
iv_main_posix.c:(.text+0x5a): undefined reference to `pthread_setspecific'
../src/.libs/libivykis.a(iv_main_posix.o): In function `iv_state_destructor':
iv_main_posix.c:(.text+0xaa): undefined reference to `pthread_setspecific'
../src/.libs/libivykis.a(iv_main_posix.o): In function `iv_init':
iv_main_posix.c:(.text+0xe3): undefined reference to `pthread_key_create'
iv_main_posix.c:(.text+0x134): undefined reference to `pthread_setspecific'
../src/.libs/libivykis.a(iv_thread_posix.o): In function `iv_thread_tls_deinit_thread':
iv_thread_posix.c:(.text+0x11b): undefined reference to `pthread_detach'
../src/.libs/libivykis.a(iv_thread_posix.o): In function `iv_thread_handler':
iv_thread_posix.c:(.text+0x229): undefined reference to `__pthread_register_cancel'
iv_thread_posix.c:(.text+0x250): undefined reference to `__pthread_unregister_cancel'
../src/.libs/libivykis.a(iv_thread_posix.o): In function `iv_thread_died':
iv_thread_posix.c:(.text+0x2c6): undefined reference to `pthread_join'
../src/.libs/libivykis.a(iv_thread_posix.o): In function `iv_thread_create':
iv_thread_posix.c:(.text+0x409): undefined reference to `pthread_create'
collect2: error: ld returned 1 exit status
```

` make CFLAGS="-Werror -pthread"` solves this but I don't want to add `-pthread` to `.travis.yml`.

I've checked the build log on my own computer, and `make` produces this
line here:

`/bin/bash ../libtool  --tag=CC   --mode=link gcc  -O0 -g -pipe -Wall
-pthread   -o iv_event_test iv_event_test.o ../src/libivykis.la`

Could you elaborate why `-pthread` is missing here?

On Wed, Dec 16, 2015 at 2:10 PM, Tibor Benke <notifications@github.com>
wrote:

> I wanted to turn this on, but I see some warnings in internal ivykis:
>
> /bin/bash ../libtool --tag=CC   --mode=link gcc  -Werror   -o iv_event_test iv_event_test.o ../src/libivykis.la
> libtool: link: gcc -Werror -o iv_event_test iv_event_test.o  ../src/.libs/libivykis.a
> ../src/.libs/libivykis.a(iv_main_posix.o): In function `__iv_deinit':
> iv_main_posix.c:(.text+0x5a): undefined reference to `pthread_setspecific'
> ../src/.libs/libivykis.a(iv_main_posix.o): In function `iv_state_destructor':
> iv_main_posix.c:(.text+0xaa): undefined reference to `pthread_setspecific'
> ../src/.libs/libivykis.a(iv_main_posix.o): In function `iv_init':
> iv_main_posix.c:(.text+0xe3): undefined reference to `pthread_key_create'
> iv_main_posix.c:(.text+0x134): undefined reference to `pthread_setspecific'
> ../src/.libs/libivykis.a(iv_thread_posix.o): In function `iv_thread_tls_deinit_thread':
> iv_thread_posix.c:(.text+0x11b): undefined reference to `pthread_detach'
> ../src/.libs/libivykis.a(iv_thread_posix.o): In function `iv_thread_handler':
> iv_thread_posix.c:(.text+0x229): undefined reference to `__pthread_register_cancel'
> iv_thread_posix.c:(.text+0x250): undefined reference to `__pthread_unregister_cancel'
> ../src/.libs/libivykis.a(iv_thread_posix.o): In function `iv_thread_died':
> iv_thread_posix.c:(.text+0x2c6): undefined reference to `pthread_join'
> ../src/.libs/libivykis.a(iv_thread_posix.o): In function `iv_thread_create':
> iv_thread_posix.c:(.text+0x409): undefined reference to `pthread_create'
> collect2: error: ld returned 1 exit status
>
> make CFLAGS="-Werror -pthread" solves this but I don't want to add
> -pthread to .travis.yml.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balabit/syslog-ng/issues/776#issuecomment-165099599>.
>

I can see -pthread on that line.
On Dec 16, 2015 5:43 PM, "Tamas Nagy" <notifications@github.com> wrote:

> I've checked the build log on my own computer, and `make` produces this
> line here:
>
> `/bin/bash ../libtool --tag=CC --mode=link gcc -O0 -g -pipe -Wall
> -pthread -o iv_event_test iv_event_test.o ../src/libivykis.la`
>
> Could you elaborate why `-pthread` is missing here?
>
> On Wed, Dec 16, 2015 at 2:10 PM, Tibor Benke <notifications@github.com>
> wrote:
>
> > I wanted to turn this on, but I see some warnings in internal ivykis:
> >
> > /bin/bash ../libtool --tag=CC --mode=link gcc -Werror -o iv_event_test
> iv_event_test.o ../src/libivykis.la
> > libtool: link: gcc -Werror -o iv_event_test iv_event_test.o
> ../src/.libs/libivykis.a
> > ../src/.libs/libivykis.a(iv_main_posix.o): In function `__iv_deinit':
> > iv_main_posix.c:(.text+0x5a): undefined reference to
> `pthread_setspecific'
> > ../src/.libs/libivykis.a(iv_main_posix.o): In function
> `iv_state_destructor':
> > iv_main_posix.c:(.text+0xaa): undefined reference to
> `pthread_setspecific'
> > ../src/.libs/libivykis.a(iv_main_posix.o): In function `iv_init':
> > iv_main_posix.c:(.text+0xe3): undefined reference to `pthread_key_create'
> > iv_main_posix.c:(.text+0x134): undefined reference to
> `pthread_setspecific'
> > ../src/.libs/libivykis.a(iv_thread_posix.o): In function
> `iv_thread_tls_deinit_thread':
> > iv_thread_posix.c:(.text+0x11b): undefined reference to `pthread_detach'
> > ../src/.libs/libivykis.a(iv_thread_posix.o): In function
> `iv_thread_handler':
> > iv_thread_posix.c:(.text+0x229): undefined reference to
> `__pthread_register_cancel'
> > iv_thread_posix.c:(.text+0x250): undefined reference to
> `__pthread_unregister_cancel'
> > ../src/.libs/libivykis.a(iv_thread_posix.o): In function
> `iv_thread_died':
> > iv_thread_posix.c:(.text+0x2c6): undefined reference to `pthread_join'
> > ../src/.libs/libivykis.a(iv_thread_posix.o): In function
> `iv_thread_create':
> > iv_thread_posix.c:(.text+0x409): undefined reference to `pthread_create'
> > collect2: error: ld returned 1 exit status
> >
> > make CFLAGS="-Werror -pthread" solves this but I don't want to add
> > -pthread to .travis.yml.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/balabit/syslog-ng/issues/776#issuecomment-165099599
> >.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balabit/syslog-ng/issues/776#issuecomment-165169838>.
>

@bazsi: Just to elaborate, do you mean on your own computer too? Do you get
any such warnings?

On Wed, Dec 16, 2015 at 9:42 PM, Balazs Scheidler <notifications@github.com>
wrote:

> I can see -pthread on that line.
>
> On Dec 16, 2015 5:43 PM, "Tamas Nagy" <notifications@github.com> wrote:
>
> > I've checked the build log on my own computer, and `make` produces this
> > line here:
> >
> > `/bin/bash ../libtool --tag=CC --mode=link gcc -O0 -g -pipe -Wall
> > -pthread -o iv_event_test iv_event_test.o ../src/libivykis.la`
> >
> > Could you elaborate why `-pthread` is missing here?
> >
> > On Wed, Dec 16, 2015 at 2:10 PM, Tibor Benke <notifications@github.com>
> > wrote:
> >
> > > I wanted to turn this on, but I see some warnings in internal ivykis:
> > >
> > > /bin/bash ../libtool --tag=CC --mode=link gcc -Werror -o iv_event_test
> > iv_event_test.o ../src/libivykis.la
> > > libtool: link: gcc -Werror -o iv_event_test iv_event_test.o
> > ../src/.libs/libivykis.a
> > > ../src/.libs/libivykis.a(iv_main_posix.o): In function `__iv_deinit':
> > > iv_main_posix.c:(.text+0x5a): undefined reference to
> > `pthread_setspecific'
> > > ../src/.libs/libivykis.a(iv_main_posix.o): In function
> > `iv_state_destructor':
> > > iv_main_posix.c:(.text+0xaa): undefined reference to
> > `pthread_setspecific'
> > > ../src/.libs/libivykis.a(iv_main_posix.o): In function `iv_init':
> > > iv_main_posix.c:(.text+0xe3): undefined reference to
> `pthread_key_create'
> > > iv_main_posix.c:(.text+0x134): undefined reference to
> > `pthread_setspecific'
> > > ../src/.libs/libivykis.a(iv_thread_posix.o): In function
> > `iv_thread_tls_deinit_thread':
> > > iv_thread_posix.c:(.text+0x11b): undefined reference to
> `pthread_detach'
> > > ../src/.libs/libivykis.a(iv_thread_posix.o): In function
> > `iv_thread_handler':
> > > iv_thread_posix.c:(.text+0x229): undefined reference to
> > `__pthread_register_cancel'
> > > iv_thread_posix.c:(.text+0x250): undefined reference to
> > `__pthread_unregister_cancel'
> > > ../src/.libs/libivykis.a(iv_thread_posix.o): In function
> > `iv_thread_died':
> > > iv_thread_posix.c:(.text+0x2c6): undefined reference to `pthread_join'
> > > ../src/.libs/libivykis.a(iv_thread_posix.o): In function
> > `iv_thread_create':
> > > iv_thread_posix.c:(.text+0x409): undefined reference to
> `pthread_create'
> > > collect2: error: ld returned 1 exit status
> > >
> > > make CFLAGS="-Werror -pthread" solves this but I don't want to add
> > > -pthread to .travis.yml.
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <
> https://github.com/balabit/syslog-ng/issues/776#issuecomment-165099599
> > >.
> > >
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/balabit/syslog-ng/issues/776#issuecomment-165169838
> >.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balabit/syslog-ng/issues/776#issuecomment-165232925>.
>

I missed the build output as it was hidden while reading your email on my
phone. Sorry for the confusion. I don't know if I have -pthread on my box
as I don't have access to my computer right now.

On the other hand I do know that ivykis used to use weak symbols on
-lpthread to make that library optional.
On Dec 17, 2015 9:08 AM, "Tamas Nagy" <notifications@github.com> wrote:

> @bazsi: Just to elaborate, do you mean on your own computer too? Do you get
> any such warnings?
>
> On Wed, Dec 16, 2015 at 9:42 PM, Balazs Scheidler <
> notifications@github.com>
> wrote:
>
> > I can see -pthread on that line.
> >
> > On Dec 16, 2015 5:43 PM, "Tamas Nagy" <notifications@github.com> wrote:
> >
> > > I've checked the build log on my own computer, and `make` produces this
> > > line here:
> > >
> > > `/bin/bash ../libtool --tag=CC --mode=link gcc -O0 -g -pipe -Wall
> > > -pthread -o iv_event_test iv_event_test.o ../src/libivykis.la`
> > >
> > > Could you elaborate why `-pthread` is missing here?
> > >
> > > On Wed, Dec 16, 2015 at 2:10 PM, Tibor Benke <notifications@github.com
> >
> > > wrote:
> > >
> > > > I wanted to turn this on, but I see some warnings in internal ivykis:
> > > >
> > > > /bin/bash ../libtool --tag=CC --mode=link gcc -Werror -o
> iv_event_test
> > > iv_event_test.o ../src/libivykis.la
> > > > libtool: link: gcc -Werror -o iv_event_test iv_event_test.o
> > > ../src/.libs/libivykis.a
> > > > ../src/.libs/libivykis.a(iv_main_posix.o): In function `__iv_deinit':
> > > > iv_main_posix.c:(.text+0x5a): undefined reference to
> > > `pthread_setspecific'
> > > > ../src/.libs/libivykis.a(iv_main_posix.o): In function
> > > `iv_state_destructor':
> > > > iv_main_posix.c:(.text+0xaa): undefined reference to
> > > `pthread_setspecific'
> > > > ../src/.libs/libivykis.a(iv_main_posix.o): In function `iv_init':
> > > > iv_main_posix.c:(.text+0xe3): undefined reference to
> > `pthread_key_create'
> > > > iv_main_posix.c:(.text+0x134): undefined reference to
> > > `pthread_setspecific'
> > > > ../src/.libs/libivykis.a(iv_thread_posix.o): In function
> > > `iv_thread_tls_deinit_thread':
> > > > iv_thread_posix.c:(.text+0x11b): undefined reference to
> > `pthread_detach'
> > > > ../src/.libs/libivykis.a(iv_thread_posix.o): In function
> > > `iv_thread_handler':
> > > > iv_thread_posix.c:(.text+0x229): undefined reference to
> > > `__pthread_register_cancel'
> > > > iv_thread_posix.c:(.text+0x250): undefined reference to
> > > `__pthread_unregister_cancel'
> > > > ../src/.libs/libivykis.a(iv_thread_posix.o): In function
> > > `iv_thread_died':
> > > > iv_thread_posix.c:(.text+0x2c6): undefined reference to
> `pthread_join'
> > > > ../src/.libs/libivykis.a(iv_thread_posix.o): In function
> > > `iv_thread_create':
> > > > iv_thread_posix.c:(.text+0x409): undefined reference to
> > `pthread_create'
> > > > collect2: error: ld returned 1 exit status
> > > >
> > > > make CFLAGS="-Werror -pthread" solves this but I don't want to add
> > > > -pthread to .travis.yml.
> > > >
> > > > —
> > > > Reply to this email directly or view it on GitHub
> > > > <
> > https://github.com/balabit/syslog-ng/issues/776#issuecomment-165099599
> > > >.
> > > >
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <
> https://github.com/balabit/syslog-ng/issues/776#issuecomment-165169838
> > >.
> >
> > >
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/balabit/syslog-ng/issues/776#issuecomment-165232925
> >.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balabit/syslog-ng/issues/776#issuecomment-165379167>.
>

--- a/.travis.yml
+++ b/.travis.yml
@@ -11,7 +11,7 @@ before_script:
   - ./autogen.sh
 script:
   - unset PYTHON_CFLAGS # HACK
-  - ./configure --with-ivykis=internal --prefix=$HOME/install/syslog-ng --enable-pacct --enable-manpages --with-docbook=/usr/share/xml/docbook/stylesheet/docbook-xsl/manpages/docbook.xsl 
+  - ./configure CFLAGS=-Werror --with-ivykis=internal --prefix=$HOME/install/syslog-ng --enable-pacct --enable-manpages --with-docbook=/usr/share/xml/docbook/stylesheet/docbook-xsl/manpages/docbook.xsl 
   - make && make distcheck VERBOSE=1 && make func-test VERBOSE=1
 compiler:
   - gcc

--- a/lib/tests/test_pathutils.c
+++ b/lib/tests/test_pathutils.c
@@ -8,13 +8,17 @@
 void
 test_is_directory_return_false_in_case_of_regular_file(void)
 {
-  int fd = open("test.file", O_CREAT, 0644);
-  write(fd, "a", 1);
-  close(fd);
+  int fd = open("test.file", O_CREAT | O_RDWR, 0644);
+  assert_false(fd < 0, "open error");
+  ssize_t done = write(fd, "a", 1);
+  assert_false(done != 1, "write error");
+  int error = close(fd);
+  assert_false(error, "close error");
 
   assert_false(is_file_directory("test.file"), "File is not a directory!");
 
-  unlink("test.file");
+  error = unlink("test.file");
+  assert_false(error, "unlink error");
 }
 
 int

--- a/modules/dbparser/pdb-rule.c
+++ b/modules/dbparser/pdb-rule.c
@@ -65,12 +65,14 @@ pdb_rule_set_context_timeout(PDBRule *self, gint timeout)
 void
 pdb_rule_set_context_scope(PDBRule *self, const gchar *scope, GError **error)
 {
-  self->context_scope = correllation_key_lookup_scope(scope);
-  if (self->context_scope < 0)
+  int context_scope = correllation_key_lookup_scope(scope);
+  if (context_scope < 0)
     {
       self->context_scope = RCS_GLOBAL;
       g_set_error(error, 0, 1, "Unknown context scope: %s", scope);
     }
+  else
+    self->context_scope = context_scope;
 }
 
 void

--- a/modules/java/native/java-class-loader.c
+++ b/modules/java/native/java-class-loader.c
@@ -25,6 +25,7 @@
 #include "java-class-loader.h"
 #include "java_machine.h"
 #include "messages.h"
+#include "lib/reloc.h"
 
 #define SYSLOG_NG_CLASS_LOADER  "org/syslog_ng/SyslogNgClassLoader"
 #define SYSLOG_NG_JAR           "syslog-ng-core.jar"

--- a/modules/java/native/java_machine.c
+++ b/modules/java/native/java_machine.c
@@ -26,6 +26,7 @@
 #include "syslog-ng.h"
 #include "messages.h"
 #include "atomic.h"
+#include "lib/reloc.h"
 
 struct _JavaVMSingleton
 {

--- a/modules/kvformat/tests/test_kv_parser.c
+++ b/modules/kvformat/tests/test_kv_parser.c
@@ -78,15 +78,6 @@ parse_kv_into_log_message(const gchar *kv)
 }
 
 static void
-assert_kv_parser_fails(const gchar *kv)
-{
-  LogMessage *msg;
-
-  msg = parse_kv_into_log_message_no_check(kv);
-  assert_null(msg, "expected kv-parser failure and it returned success, kv=%s", kv);
-}
-
-static void
 test_kv_parser_basics(void)
 {
   LogMessage *msg;

--- a/modules/python/python-dest.h
+++ b/modules/python/python-dest.h
@@ -25,6 +25,7 @@
 #ifndef _SNG_PYTHON_DEST_H
 #define _SNG_PYTHON_DEST_H
 
+#include "python-module.h"
 #include "driver.h"
 #include "logwriter.h"
 #include "value-pairs.h"

--- a/modules/python/python-module.h
+++ b/modules/python/python-module.h
@@ -24,7 +24,45 @@
 #ifndef PYTHON_MODULE_H_INCLUDED
 #define PYTHON_MODULE_H_INCLUDED 1
 
-#include "syslog-ng.h"
+/*
+ * There are some Python related things to keep in mind when working on the
+ * Python module.  The Python header unfortunately defines a few standard
+ * related conditionals (_POSIX_C_SOURCE and _XOPEN_SOURCE) in Python.h that
+ * it shouldn't, causing two issues potentially:
+ *
+ *   1) warnings, as these defines are usually defined by our configure
+ *      script/config.h as well
+ *
+ *   2) triggers POSIX and XOPEN specific interfaces/incompatibilies to be
+ *      enabled in system headers.  This could cause incompatibilities
+ *      between C code that include Python.h and C code that doesn't.
+ *      Imagine the case when a new struct member is added based on the
+ *      preprocessor defines above.  In one part of the code that member
+ *      wouldn't exist, in others it would.
+ *
+ *      Fortunately this rarely causes issues in practice for a number of reasons:
+ *        a) syslog-ng is compiled with the same as _GNU_SOURCE is defined
+ *           that implies the other two.<
+ *
+ *        b) system libraries are usually designed in a way that struct members
+ *           are never removed/added but rather their name is changed (for
+ *           example by prefixing them with a pair of underscores)
+ *
+ * Our workaround here is to check whether Python.h inclusion comes first
+ * and if it's not we error out during compilation.  Within the Python
+ * module of syslog-ng we should keep the convention to make sure that
+ * python-module.h is the first included file in any C modules.
+ *
+ * python-module.h also makes sure that syslog-ng.h is included (so that
+ * proper ordering happens), so you can omit syslog-ng.h from all
+ * python-module headers.
+ */
+
+#if defined(_POSIX_C_SOURCE) || defined(_XOPEN_SOURCE)
+#error "_POSIX_C_SOURCE or _XOPEN_SOURCE is already defined, python-module.h should be included first. Check out the comment in python-module.h for more information"
+#endif
+
 #include <Python.h>
+#include "syslog-ng.h"
 
 #endif

--- a/modules/python/python-parser.h
+++ b/modules/python/python-parser.h
@@ -25,9 +25,10 @@
 #ifndef SNG_PYTHON_PARSER_H_INCLUDED
 #define SNG_PYTHON_PARSER_H_INCLUDED
 
+#include "python-module.h"
+#include "driver.h"
 #include "cfg-parser.h"
 #include "cfg-lexer.h"
-#include "python-dest.h"
 
 extern CfgParser python_parser;
 

--- a/modules/python/python-tf.c
+++ b/modules/python/python-tf.c
@@ -22,7 +22,6 @@
  *
  */
 #include "python-tf.h"
-#include "python-module.h"
 #include "python-helpers.h"
 #include "python-main.h"
 #include "template/simple-function.h"

--- a/modules/python/python-tf.h
+++ b/modules/python/python-tf.h
@@ -24,6 +24,7 @@
 #ifndef _SNG_PYTHON_TF_H
 #define _SNG_PYTHON_TF_H
 
+#include "python-module.h"
 #include "lib/template/function.h"
 
 TEMPLATE_FUNCTION_PROTOTYPE(tf_python);

--- a/tests/loggen/loggen.c
+++ b/tests/loggen/loggen.c
@@ -330,7 +330,7 @@ format_timezone_offset_with_colon(char *timestamp, int timestamp_size, struct tm
   offset[len - 1] = offset[len - 2];
   offset[len - 2] = ':';
 
-  strncat(timestamp, offset, sizeof(offset));
+  strncat(timestamp, offset, timestamp_size - strlen(timestamp) -1);
 }
 
 static guint64

